### PR TITLE
feat: add sample method to document array

### PR DESCRIPTION
**Pull request by bwanglzu**
add `sample` to DocumentArray

### CODE REVIEW

File: jina/types/arrays/document.py [hunk 1]
```diff
@@ -445,3 +446,16 @@ def load_binary(cls, file: Union[str, BinaryIO]) -> 'DocumentArray':
...
+
+    def sample(self, k: int) -> 'DocumentArray':
+        """random sample k elements from :class:`DocumentArray` without replacement
+
+        :param k: Number of elements to sample.
+        """
+        if k < len(self):
+            raise ValueError('DocumentArray do not have enough Document to sample')
+        indices = random.sample(range(len(self)), k)
+        sampled = DocumentArray()
```
Comment (JoanFM) [inline jina/types/arrays/document.py:458]: I think u can do `sampled = self[indices]`

Comment (bwanglzu) [reply jina/types/arrays/document.py]: it's not woking

Comment (JoanFM) [reply jina/types/arrays/document.py]: why not? we should maybe understand why first?

Comment (bwanglzu) [reply jina/types/arrays/document.py]: @JoanFM it's not a python syntax, it's supported by numpy only

File: jina/types/arrays/search_ops.py [hunk 2]
```diff
@@ -73,3 +74,21 @@ def find(
...
+
+        :param k: Number of elements to sample from the document array.
+        :param seed: initialize the random number generator, by default is None. If set will
+            save the state of the random function to produce certain outputs.
+        :return: A sampled list of :class:`Document` represented as :class:`DocumentArray`.
+        """
+        from .document import DocumentArray
+
+        if k > len(self):
+            raise ValueError('DocumentArray do not have enough Document to sample')
```
Comment (hanxiao) [inline jina/types/arrays/search_ops.py:89]: better tell user how many in DA are k must be smaller than this number

File: jina/types/arrays/search_ops.py [hunk 3]
```diff
@@ -73,3 +74,21 @@ def find(
...
+        if seed:
```
Comment (hanxiao) [inline jina/types/arrays/search_ops.py:90]: seed =0 wont trigger this line, so has to be:

```suggestion
        if seed is not None:
```

Comment (bwanglzu) [reply jina/types/arrays/search_ops.py]: fixed

File: tests/unit/types/arrays/test_documentarray.py [hunk 4]
```diff
@@ -397,3 +397,17 @@ def test_cache_invalidation_sort_reverse(docarray_for_cache):
...
+def test_sample():
+    da = DocumentArray(random_docs(100))
+    sampled = da.sample(5)
+    assert len(sampled) == 5
+    assert isinstance(sampled, DocumentArray)
+    sampled_seed_1 = da.sample(5, seed=1)
+    sampled_seed_2 = da.sample(5, seed=1)
+    assert len(sampled_seed_1) == 5
+    assert sampled_seed_1 == sampled_seed_2
+    assert sampled != sampled_seed_1
```
Comment (tadejsv) [inline tests/unit/types/arrays/test_documentarray.py:411]: This can break.

File: tests/unit/types/arrays/test_documentarray.py [hunk 5]
```diff
@@ -397,3 +397,17 @@ def test_cache_invalidation_sort_reverse(docarray_for_cache):
     docarray_for_cache.reverse()
     assert docarray_for_cache[0].id == '2'
     assert docarray_for_cache[1].id == '1'
+
+
+def test_sample():
```
Comment (tadejsv) [inline tests/unit/types/arrays/test_documentarray.py:402]: I think you can break this unit test into two: one testing that correct amount of items are returned, another testing that setting random seed works as expected

Comment (bwanglzu) [reply tests/unit/types/arrays/test_documentarray.py:402]: true, PR updated

File: .github/2.0/cookbooks/Document.md [hunk 6]
```diff
@@ -1149,6 +1163,25 @@ for d in dam:
...
+from jina import Document
+from jina.types.arrays.memmap import DocumentArrayMemmap
+
+dam = DocumentArrayMemmap('./my-memmap')  # initialize a random document array
+for idx in range(100):
+    dam.append(Document(id=idx))  # append 100 documents into `da`
+sampled_dam = dam.sample(k=10)  # sample 10 documents
+sampled_dam_with_seed = dam.sample(k=10, seed=1)  # sample 10 documents with seed.
+```
+
```
Comment (hanxiao) [inline .github/2.0/cookbooks/Document.md:1184]:
```suggestion

```

**Review by hanxiao — APPROVED** [review]

**Pull request merged**